core/domain/state_domain.py

```diff
@@ -916,8 +938,946 @@ def get_all_outcomes(self) -> List[Outcome]:
+        )
+
+    def _get_rule_value_of_fraction_interaction(
+        self,
+        rule_spec: RuleSpec
+    ) -> bool:
+        """Return rule value of the rule_spec of FractionInput interaction so
+        that we can keep track of rule's range.
+
+        Args:
+            rule_spec: RuleSpec. Rule spec of an answer group.
+
+        Returns:
+            rule_value_f: float. The value of the rule spec.
+        """
+        rule_value_f = rule_spec.inputs['f']
+        rule_value_f = (
+            rule_value_f['wholeNumber'] +
+            float(rule_value_f['numerator']) / rule_value_f['denominator']
+        )
+
+        return rule_value_f
+
+    def _validate_numeric_input(self) -> None:
+        """Validates the NumericInput interaction.
+
+        Raises:
+            ValidationError. Raises a validation error in the following
+                condition:
+                - Duplicate rules are present.
+                - Rule having a solution that is subset of previous rules'
+                solution.
+                - `tol` value in `IsWithinTolerance` is negetive.
+                - `a` is greater than or equal to `b` in `IsInclusivelyBetween`
+                rule.
+        """
+        lower_infinity = float('-inf')
+        upper_infinity = float('inf')
+        ranges: List[InteractionInstance.RangeVariableDict] = []
+        rule_spec_till_now: List[RuleSpecDict] = []
+
+        for ans_group_index, answer_group in enumerate(self.answer_groups):
+            for rule_spec_index, rule_spec in enumerate(
+                answer_group.rule_specs
+            ):
+                # Rule should not be duplicate.
+                if rule_spec.to_dict() in rule_spec_till_now:
+                    raise utils.ValidationError(
+                        f'The rule \'{rule_spec_index}\' of answer group '
+                        f'\'{ans_group_index}\' of NumericInput '
+                        f'interaction is already present.'
+                    )
+                rule_spec_till_now.append(rule_spec.to_dict())
+                # All rules should have solutions that is not subset of
+                # previous rules' solutions.
+                range_var: InteractionInstance.RangeVariableDict = {
+                    'ans_group_index': int(ans_group_index),
+                    'rule_spec_index': int(rule_spec_index),
+                    'lower_bound': None,
+                    'upper_bound': None,
+                    'lb_inclusive': False,
+                    'ub_inclusive': False
+                }
+
+                if rule_spec.rule_type == 'IsLessThanOrEqualTo':
+                    rule_value = float(rule_spec.inputs['x'])
+                    self._set_lower_and_upper_bounds(
+                        range_var, lower_infinity,
+                        rule_value, False, True
+                    )
+
+                if rule_spec.rule_type == 'IsGreaterThanOrEqualTo':
```

Comment:
Should all these if be rather elif block?

Reply:
Done